R/Model-class.R

```diff
@@ -2156,189 +2156,174 @@ Effloglog <- function(eff,
+    rw1 = "logical",
+    X = "matrix",
+    RW = "matrix",
+    RW_rank = "integer"
+  ),
+  prototype = prototype(
+    eff = c(0, 0),
+    eff_dose = c(1, 1),
+    sigma2W = 0.025,
+    sigma2betaW = 1,
+    rw1 = TRUE,
+    use_fixed = c(sigma2W = TRUE, sigma2betaW = TRUE)
+  ),
+  contains = "ModelEff",
+  validity = v_model_eff_flexi
 )
-{##No observed Efficacy response
-  if (length(data@w)==0){
-    w1<-Eff
-    x1<-Effdose} else {
-      ## with observed efficacy responses and no DLT observed
-      w1<-c(Eff, getEff(data)$w_no_dlt)
-      x1<-c(Effdose, getEff(data)$x_no_dlt)
-    }
-  ## Match dose levels in x1 with the all dose levels for evaluations
-  x1Level <- matchTolerance(x1,data@doseGrid)
-  smooth<-match.arg(smooth)
-  useRW1<- smooth == "RW1"
-  useFixed<-list()
-  for (parName in c("sigma2","sigma2betaW"))
-  {useFixed[[parName]] <- identical(length(get(parName)),1L)}
-  #design matrics
-  designW <- model.matrix(~ -1 + I(factor(x1Level, levels=seq_len(data@nGrid))))
-  dimnames(designW) <- list(NULL,NULL)
-
-  ##difference matrix of order 1:
-  D1mat<- cbind(0,diag(data@nGrid-1)) - cbind(diag(data@nGrid - 1),0)
-
-  ## set up the random walk penalty matrix and its rank:
-  if (useRW1)
-  {## the rank-deficient prior precision for the RW1 prior:
-    RWmat <- crossprod(D1mat)
-    ##Rank: dimension -1
-    RWmatRank <- data@nGrid-1L
-  } else {##second-order difference
-    D2mat <- D1mat[-1,-1] %*% D1mat
-    RWmat <- crossprod(D2mat)
-    RWmatRank <- data@nGrid-2L
+
+## constructor ----
+
+#' @rdname EffFlexi-class
+#'
+#' @param eff (`numeric`)\cr the pseudo efficacy responses.
+#'   Elements of `eff` must correspond to the elements of `eff_dose`.
+#' @param eff_dose (`numeric`)\cr dose levels that correspond to pseudo efficacy
+#'   responses in `eff`.
+#' @param sigma2W (`numeric`)\cr the prior variance of the efficacy responses.
+#'   This is either a fixed value or a named vector with two positive numbers,
+#'   the shape (`a`), and the rate (`b`) parameters for the inverse gamma
+#'   distribution.
+#' @param sigma2betaW (`numeric`)\cr the prior variance of the random walk model
+#'   used for smoothing. This is either a fixed value or a named vector with two
+#'   positive numbers, the shape (`a`), and the rate (`b`) parameters for the
+#'   inverse gamma distribution.
+#' @param rw1 (`flag`)\cr used for smoothing data for this efficacy model. If it
+#'   is `TRUE`, the first-order random walk model is used for the mean efficacy
+#'   responses. Otherwise, the random walk of second order is used.
+#' @param data (`DataDual`)\cr observed data to update estimates of the model
+#'   parameters.
+#'
+#' @export
+#' @example examples/Model-class-EffFlexi.R
+#'
+EffFlexi <- function(eff,
+                     eff_dose,
+                     sigma2W,
+                     sigma2betaW,
+                     rw1 = TRUE,
+                     data) {
+  assert_numeric(eff)
+  assert_numeric(eff_dose)
+  assert_flag(rw1)
+
```

Comment:
also need to assert `data`, potentially also `sigma2W` and `sigma2betaW`

Reply:
Tx @danielinteractive, you mean `assert_true(validObject(data))`for `data`?
I am wondering about `sigma2W` and `sigma2betaW` (and maybe also about `data`), as we do no not use it much in the bodu of the user constructor, and these objects are later validated by the class validation function. What do you think?

Reply:
I would just say `assert_class(data, "Data")` or what the class should be.
For the sigmas, as soon as we touch it within the constructor, then I would assert it (just `numeric`). otherwise not.